Retry functionality

Auwate opened the pull request:
# Description

This PR seeks to resolve issue #8, which raises the issue of retrying on cluster unavailability. This PR uses tenacity and it's accompanying retry feature to solve this.

Auwate committed:
Commit: Adding /docs/blueprint

# Description:
To progress the project, a blueprint needs to be built to determine the direction of the project. This way, it can be formally accepted by the Kytos team.

# Modifications:
- Added /docs/blueprint/EP001.rst

Auwate committed:
Commit: EP039

# Description:
- Added docs/blueprint
  - Added EP039.rst

Auwate committed:
Commit:
# Description:
- Working on issue 8

Auwate committed:
Commit:

# Description: Adding relevant files for CI pipeline

Auwate committed:
Commit:

# Fixing CI issues

Auwate committed:
Commit:

# Getting testing to work

Auwate committed:
Commit:

# Description:
- Refactoring relative path

viniarck commented on main.py:
```diff
@@ -213,14 +269,21 @@ class AsyncScheduler:
     """
 
-    def __init__(self, loop: asyncio.AbstractEventLoop = None, thread: Thread = None, coroutine: Coroutine = None):
+    def __init__(
```
Let's not forget this discussion:

https://github.com/amlight/kafka_napp/pull/7/files#r1972165225

No need to address here. Also, side comment: I'll recommend that you finish the blueprint and get it accepted. Finally, when it comes the time to move this repo to `kytos-ng` all the code (and previous committed) will be reviewed again, in the future I'll ask you do to an initial commit there, and then push all the code. Points that were raised are expected to be resolved by then.

Auwate replied:
Hey Vinicius, good morning. I understand, I was under the impression that the blueprint outlined what the NApp should already be able to do (essentially providing a minimal viable product) that can be expanded. In doing so, I wanted to update parts of it, as there are things I could change. This was one of those features.

Nonetheless, I'll focus on getting the blueprint accepted after this PR and working on implementing the portions we agreed on for the blueprint.

Auwate replied:
Got it. I'm still looking at setting up a benchmark test to see the tradeoff in resource consumption to output with the added layer of complexity.

I'll focus on getting the blueprint accepted for the future though, thanks for bringing this up.

viniarck commented on requirements/dev.in:
```diff
@@ -0,0 +1,10 @@
+
+-r run.in
+
+pylint==3.1.0
```
`pylint`, `pytest` and `black` should come from `-e git+https://github.com/kytos-ng/kytos.git#egg=kytos[dev]` https://github.com/kytos-ng/kytos/blob/master/setup.py#L224-L236

Only the current first 4 lines are needed here. Otherwise, this will make harder to maintain these deps. NApps from kytos-ng are expected to follow our current testing dependency patterns and how they're pinned.

Auwate replied:
Ah I see, thank you for clarifying. I'll remove these

viniarck commented on requirements/run.in:
```diff
@@ -0,0 +1,5 @@
+aiokafka==0.12.0
+kafka-python-ng==2.2.3
```
kafka-python-ng isn't `asyncio` oriented, any new NApp we're building should be compatible and use `asyncio` well maintained libraries. Did you hit anything else missing with `aiokafka`? 

`kafka-python-ng` it's even rolling out their own `Future`s

- https://github.com/kafka-python-ng/kafka-python-ng/blob/master/kafka/admin/client.py#L372
- https://github.com/kafka-python-ng/kafka-python-ng/blob/master/kafka/future.py

I don't look forward to maintain this unless we have a critical reason to go for it.

Auwate replied:
Yes `kafka-python-ng` is not `asyncio` oriented but it provides an admin class that can help create a topic if one does not exist. It can be removed but this would require a manual topic creation if the cluster metadata is wiped or the topic name needs to be changed.

I'm okay with the removal as it reduces dependencies, but it adds some complexity that should be mentioned in the blueprint and README.

viniarck replied:
Right. Let's remove it then. Dependencies can be a liability too. 

Let's stick with asyncio oriented and well maintained ones, especially on new NApps like this one.

viniarck commented on requirements/run.in:
```diff
@@ -0,0 +1,5 @@
+aiokafka==0.12.0
+kafka-python-ng==2.2.3
+tenacity==8.2.3
```
`tenacity` comes from `kytos` deps, so no need to add here again, `packaging` and `typing-extensions` can be removed.

viniarck commented on setup.cfg:
```diff
@@ -0,0 +1,23 @@
+[yala]
+radon mi args = --min C
+pylint args = --disable=too-many-locals,too-few-public-methods,too-many-instance-attributes,too-many-arguments,inconsistent-return-statements,unnecessary-pass,too-many-public-methods,redefined-outer-name,unnecessary-lambda,missing-timeout,import-error,no-name-in-module,attribute-defined-outside-init --ignored-modules=napps.kytos.of_lldp
+linters=pylint,pycodestyle,isort
```
Newer NApps like this one should also have `black`. Can you include it?

viniarck commented on setup.cfg:
```diff
@@ -0,0 +1,23 @@
+
+[yala]
+radon mi args = --min C
+pylint args = --disable=too-many-locals,too-few-public-methods,too-many-instance-attributes,too-many-arguments,inconsistent-return-statements,unnecessary-pass,too-many-public-methods,redefined-outer-name,unnecessary-lambda,missing-timeout,import-error,no-name-in-module,attribute-defined-outside-init --ignored-modules=napps.kytos.of_lldp
```
`of_lldp` stuff shouldn't be here, probably accidental from when you initially based on `of_lldp` config, please review this, clean it up, and also consider removing certain ignored args here if they aren't needed too.

viniarck commented on tox.ini:
```diff
@@ -0,0 +1,23 @@
+[tox]
```
Good. Once we have this NApp ready to be moved to `kytos-ng` org we'll hook it with Scrutinizer which will run the tox file, and then on each PR we'll have an automatic execution. Nice.

viniarck commented on main.py:
```diff
@@ -188,12 +213,34 @@ async def check_for_topic(self, topic_name: str) -> bool:
+            log.error("AIOKafka retrying shutdown procedure...")
+            raise exc
 
+    @retry(
```
That's a great start to also have the retries. 

Notice that this also opens up the likelihood of events out of order, imagine you have the following events:

- a) `.*.switch.interface.link_up` at t0
- b) `.*.switch.interface.link_down` at t1

Imagine that the producer temporarily fails, and then retries kicks in, while it's waiting, b) arrives at t2 and then it succeeds immediately and gets send, then a) during the retry succeeds at t3, so in the end you'd end up with producing link_down at t3 and link_up at t4. 

This is a corner case, and generic problem that any chronological or stateful events can experience. So for sending retries, we need to have a clear and coherent lock strategy too, see what I mean? Otherwise this can lead to other problems for consumers. We definitely don't want to kill the concurrency too, so this needs to be carefully thought to work well with asyncio while ensuring great throughput.

Auwate replied:
Ah I see, thank you for your input. This is an important edge case to consider.

A possible solution is to add an internal queue where failed sends are inserted in chronological order. This means that processing happens in the order they were added. However, this does increase memory by an order of N, where N is the amount of missed sends. Possible options include Python's `collection.deque` and `queue.Queue`.

If you have any thoughts that would be greatly appreciated.

Auwate replied:
I had an idea to tackle this while researching `aiokafka`. Essentially, `aiokafka` and other Python/Kafka frameworks send their data in batches to improve performance and throughput of their systems. This means that there already is a level of redundancy to protect against connections errors. However, this only comes into play if the `acks` setting is set to `1` or `all` (where `1` tells `aiokafka` to delete when the topic leader confirms, and `all` is when all kafka replicas confirm, including the topic leader). Currently we have it at `0`, which provides the highest throughput with zero latency but has no guarantees.

If we put `acks` to 1 then we have a level of redundancy built in to `aiokafka` to avoid this issue. We would still need to implement an queue for messages that can't be put into the buffer though (like if the buffer is full), so it's not 100% effective.

viniarck replied:
@Auwate

Yes, let's let's leverage `aiokafka` retries, and only implement at the application level what's absolutely needed, the more we can leverage it and Kafka built-in functionalities the better:

> aiokafka will retry most errors automatically, but only until request_timeout_ms. 

https://aiokafka.readthedocs.io/en/stable/producer.html#retries-and-message-acknowledgement 

I see that it mentions "most" errors, we gotta figure out which ones, and understand what else we need to do at the application level in the worst case. Continue with your research.

Also, yes for `acks=all` or `ack=1`, we definitely need to make a tradeoff prioritizing correctness in favor of absolutely throughput, since sending non ordered events can lead to critical logic error for consumers.

viniarck replied:
> A possible solution is to add an internal queue where failed sends are inserted in chronological order. This means that processing happens in the order they were added. However, this does increase memory by an order of N

Using an internal queue would be fine, it's OK to pay this O(n) space complexity, and we could let it grow during full outrages. However, there's still other intermediary reliable and easier to maintain approaches to consider:

- 1) Initially, let's go with just a global asyncio lock for all sending operations with `await self._send_lock:`, this will serialize the send operations (and then we'd let `aiokafka` do its job batching as much as it can), we wouldn't be leveraging its to the full potential, but just as a starting point it's a valuable exercise, if this perform good enough we can start with it, and then evolve it. Prorotype and run some tests, let's see how the throughput would look like.
- 2) You could implement a queue too since the events by default are created in order (and kytos event bus already perform a prequeue), you can leverage a [`asyncio.Event`](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Event), which will signal if any retry is going on, if any retry is going on then any any send task will wait for it (and suspends). But for this to work you would also need to use tenacity in line instead of the decorator with `async for attempt in AsyncRetrying(retry=retry_if_result(...)):`, and then whenever it starts retrying you just set the event, and from that point forward, any other task will wait for the retries to wait before keep sending.

Finally, I believe it's also worth considering enabling [produce idempotency](https://aiokafka.readthedocs.io/en/stable/producer.html#idempotent-produce), check it out and see if we can use it.

Explore these ones out and then we can keep brainstorming, and potentially consider another queue if absolutely needed, but I think we can get away without it while still shipping something that's solid in terms of reliability. And then at the application level let's only handle retries that are absolutely needed as we discussed in the prior comment.

viniarck reviewed (changes requested):
@Auwate, great progress, overall looks good, but there are things that need to be changed to be accepted. Check out my comments. Also, for future PRs, let's go for the blueprint first, and then you focus on finishing it, and get it accepted, and then you keep building and shipping the rest of the funcionalities, that will make the code review process easy too.

Auwate committed:
Commit:

# Description:
- Removed kafka-python-ng
- Updated run.in and dev.in
- Updated settings.py to have acks=1
- Updated main.py to remove KafkaAdmin

Auwate committed:
Commit:

# Description:
- Align changes with only what is necessary (no formatting bloat)
- Add lock mechanism

Auwate committed:
Commit:

# Description:
- Remove helper.py format bloat

Alopalao commented on tox.ini:
```diff
@@ -0,0 +1,23 @@
+usedevelop=True  # Use `setup.py develop` instead of `bdist_wheel`
+
+deps=
+    -rrequirements/dev.txt
```
Use `dev.in` instead

Alopalao commented on .gitignore:
```diff
@@ -0,0 +1 @@
+.eggs
```
Since I am seeing the tests starting to be created. You may want to copy the `.gitignore` from other NApps. Once you ran `tox` a bunch of files will be created.

Alopalao commented on kytos.json:
```diff
@@ -1,7 +1,7 @@
+    "name": "kafka_napp",
     "description": "Integrates with Kafka to push events and logs",
     "version": "1.0",
     "napp_dependencies": ["of_core", "topology"],
```
It looks like this NApp has no dependencies. You can an empty list here.

Alopalao commented on tox.ini:
```diff
@@ -0,0 +1,23 @@
+deps=
+    -rrequirements/dev.txt
+
+[testenv:coverage]
```
Add the option `skip_install = true` so you can start running tox.

Alopalao commented on setup.cfg:
```diff
@@ -0,0 +1,23 @@
+
+[yala]
+radon mi args = --min C
+pylint args = --disable=too-many-locals,too-few-public-methods,too-many-instance-attributes,too-many-arguments,inconsistent-return-statements,unnecessary-pass,too-many-public-methods,redefined-outer-name,unnecessary-lambda,missing-timeout,import-error,no-name-in-module,attribute-defined-outside-init
```
Now with the running `tox` give a second look at all these disables. I could be wrong but I think you do not need `missing-timeout`

Alopalao reviewed (commented):
Reading the comments, it seems that you are doing some tests for the retrying methods that we are going to be using. Otherwise, I looked at tests and see if this NApp is ready to be incorporated to kytos environment. It is pretty close, great job. Do not forget that you still need a blueprint approved.

Alopalao commented on tests/unit/test_kafka_send_ops.py:
```diff
@@ -0,0 +1,61 @@
+
+@pytest.mark.asyncio
+@patch("kafka_napp.main.AIOKafkaProducer.start")
+async def test_setup_producer_retries(start_mock):
```
You can mock `time.sleep` as well so in the tests it does not wait the 4 to 9 seconds that it would while running. For example we did the same for [sdntrace_cp](https://github.com/kytos-ng/sdntrace_cp/blob/15b8611425e242ecf92d641bc2756660aec9aab1/tests/unit/test_utils.py#L28-L35)

Auwate replied:
Thanks for the feedback! I'm going to get through each of these and get the tests working. I believe that mocking the `time.sleep` function would also really benefit the tests, so thank you for the suggestion

Auwate committed:
Commit: Add chronological guarantees

# Description:
- KafkaSendOperations testing
  - Added required setup.py and configuration files to run testing
- Added asyncio.Lock

Auwate committed:
Commit:

# Description:
- Fixed potential issue with call_soon_threadsafe not accepting a coroutine by using run_coroutine
- Added a check before sending data to confirm the producer is ready

Auwate committed:
Commit:

# Description:
- Fixed issues that was not allowing the application to work in production

Auwate commented:
# Documentation

This comment aims to discuss the recent changes that happened in this PR, as well as comments & concerns for the future that may affect this application.

## New changes

- settings.py
  - Added ENABLE_ITEMPOTENCE
  - Changed ACKS to "all"
  - Added ALLOWED_RETRIES
  - Added REQUEST_TIMEOUT_MS

- KafkaSendOperations (main.py)
  - Added features to provide some guarantees on message delivery (assuming Kafka does not stay down for too long)
    - Added asyncio.Lock to make messages wait in line until the message before them was able to be sent
    - Added retry mechanics, especially on the asyncio.Lock wrapper function
  - Changed start up sequence to use run_coroutine instead of call_callable_soon as there may have been possibilities of misuse of coroutines/callables
  - Removed kafka_python_ng dependencies and removed feature to create topics on demand
  - Added check to assert that the producer is ready to send messages to Kafka

- AsyncScheduler (main.py) (soon to be removed in favor of using MainThread)
  - Removed run_callable_soon, run_coroutine_soon_and_wait, and cancel_all_tasks as they were not going to be used
  - Added timeout for joining the thread

- requirements/
  - Added dev.in, run.in, dev.txt, and run.txt

- Added necessary files
  - kytos.json, setup.cfg, MANIFEST.in, .gitignore

- Added tox
  - tox.ini, pytest.ini

- Added testing to ensure message delivery guarantee behavior
  - Found in test_kafka_send_ops.py

## Concerns

Benchmarks between using MainThread and a separate thread came back and there showed marginal gains in throughput but at a somewhat higher CPU utilization and negligible memory consumption. This will be expressed more thoroughly at the blueprint PR.

viniarck reviewed (approved):
@Auwate great progress, nice improvements and I'm glad that you're collaborating with @Alopalao too. I appreciated also the summarized changes, overall, I think we're all on the same page and moving int the right direction (the blueprint here I didn't review and I'll review it when you send it to [kytos-ng/kytos](https://github.com/kytos-ng/kytos/tree/master/docs/blueprints)). So, let's keep moving forward, this PR can land as it is to also unblock you, go ahead and merge it.

Next steps and tasks recommendation to facilitate for you and everybody involved in the code reviews too:

1) Augment, review and send the official blueprint to [kytos-ng/kytos](https://github.com/kytos-ng/kytos/tree/master/docs/blueprints). Once it's accepted there, it'll be your and everybody's else reference when it comes to understanding what's expected in core reviews and what's being delivered. 
2) Once the blueprint is accepted, I'll create a new repo for you to continue your work on `kytos-ng` org. 
2.1) I'll also enable [Scrutinizer CI](https://scrutinizer-ci.com/) to run in the repo too, so every subsequent PR will run all unit tests and linters
2.2) Send a new PR with the all code you already have, we're not just not do a ownership transfer, that way it can also serve as a checkpoint for core developers to read the blueprint and catch up with the current progress. 
2.3) From that point forward, you can continue to break down your tasks in the new repo and keep sending PRs.


OK? Congrats on this NApp, it's shaping up well.
